redis/redis.go

```diff
@@ -283,6 +287,55 @@ func (c *RedisStore[V]) ClearAll(ctx context.Context) error {
 	return fmt.Errorf("cachestore/redis: unsupported")
 }
 
+func (c *RedisStore[V]) GetOrSetWithLock(
+	ctx context.Context, key string, getter func(context.Context, string) (V, error),
+) (V, error) {
+	var out V
+
+	mu, err := c.newMutex(ctx, key)
+	if err != nil {
+		return out, fmt.Errorf("cachestore/redis: creating mutex failed: %w", err)
+	}
+	defer mu.Unlock()
+
+	for i := 0; ; i++ {
+		// If there's a value in the cache, return it immediately
+		out, found, err := c.Get(ctx, key)
```

Comment:
I wonder if there is a race here..? and if we should acquire the lock before the get..?

Reply:
Two processes might be able to both `Get` but if it's empty then only one of them will be able to acquire the lock for writing (with `TryLock`), the other one will have to wait. Then, if the first process has already finished and set the value, the second one will call `Get` again to find a value there and exit the loop. So I don't think there's a race condition here...

If we acquire the lock before the Get then the mutex would protect the reads as well, so there'd only be a single concurrent reader